Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: PatternVariableName

child of #14942 
I have read check documentation: https://checkstyle.org/checks/naming/patternvariablename.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
record Point(int x, int y) { }
record square(Point upperLeft, Point lowerRight) { }

public class PatternVariables {
    void method(Object obj) {
        if (obj instanceof square(Point(int _, int y), _)) {  // violation
            System.out.println(y);
        }
    }
}
PS D:\CS\test> cat config.xml                                                                                
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="PatternVariableName">
    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/PatternVariables.java                
Starting audit...
[ERROR] D:\CS\test\src\PatternVariables.java:6:45: Name '_' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
Audit done.
Checkstyle ends with 1 errors.

```
---

**Describe what you expect in detail.**

`PatternVariableName` should allow unnamed pattern variables by default

---

--- a/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionPatternVariableNameTest.java
+++ b/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionPatternVariableNameTest.java
@@ -46,7 +46,7 @@ public class XpathRegressionPatternVariableNameTest extends AbstractXpathTestSup
 
         final DefaultConfiguration moduleConfig =
                 createModuleConfig(PatternVariableNameCheck.class);
-        final String defaultPattern = "^[a-z][a-zA-Z0-9]*$";
+        final String defaultPattern = "^([a-z][a-zA-Z0-9]*|_)$";
 
         final String[] expectedViolation = {
             "6:33: " + getCheckMessage(PatternVariableNameCheck.class,

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheck.java
@@ -30,7 +30,7 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * <li>
  * Property {@code format} - Sets the pattern to match valid identifiers.
  * Type is {@code java.util.regex.Pattern}.
- * Default value is {@code "^[a-z][a-zA-Z0-9]*$"}.
+ * Default value is {@code "^([a-z][a-zA-Z0-9]*|_)$"}.
  * </li>
  * </ul>
  * <p>
@@ -51,7 +51,7 @@ public class PatternVariableNameCheck extends AbstractNameCheck {
 
     /** Creates a new {@code PatternVariableNameCheck} instance. */
     public PatternVariableNameCheck() {
-        super("^[a-z][a-zA-Z0-9]*$");
+        super("^([a-z][a-zA-Z0-9]*|_)$");
     }
 
     @Override

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/PatternVariableNameCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/PatternVariableNameCheck.xml
@@ -8,7 +8,7 @@
  Checks that pattern variable names conform to a specified pattern.
  &lt;/p&gt;</description>
          <properties>
-            <property default-value="^[a-z][a-zA-Z0-9]*$"
+            <property default-value="^([a-z][a-zA-Z0-9]*|_)$"
                       name="format"
                       type="java.util.regex.Pattern">
                <description>Sets the pattern to match valid identifiers.</description>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckTest.java
@@ -48,7 +48,7 @@ public class PatternVariableNameCheckTest
     @Test
     public void testDefault() throws Exception {
 
-        final String pattern = "^[a-z][a-zA-Z0-9]*$";
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
 
         final String[] expected = {
             "18:39: " + getCheckMessage(MSG_INVALID_PATTERN, "OTHER", pattern),
@@ -96,6 +96,28 @@ public class PatternVariableNameCheckTest
         verifyWithInlineConfigParser(
                 getNonCompilablePath(
                         "InputPatternVariableNameEnhancedInstanceofNoSingleChar.java"),
+                expected);
+    }
+
+    @Test
+    public void testPatternVariableNameUnnamed() throws Exception {
+
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
+
+        final String[] expected = {
+            "17:33: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "19:33: " + getCheckMessage(MSG_INVALID_PATTERN, "_s", pattern),
+            "22:33: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "29:25: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "32:25: " + getCheckMessage(MSG_INVALID_PATTERN, "_s", pattern),
+            "40:67: " + getCheckMessage(MSG_INVALID_PATTERN, "_Color", pattern),
+            "45:59: " + getCheckMessage(MSG_INVALID_PATTERN, "_Color", pattern),
+            "51:76: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputPatternVariableNameUnnamed.java"),
                 expected);
     }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/InputPatternVariableNameEnhancedInstanceofTestDefault.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/InputPatternVariableNameEnhancedInstanceofTestDefault.java
@@ -1,6 +1,6 @@
 /*
 PatternVariableName
-format = (default)^[a-z][a-zA-Z0-9]*$
+format = (default)^([a-z][a-zA-Z0-9]*|_)$
 
 
 */

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/InputPatternVariableNameUnnamed.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/InputPatternVariableNameUnnamed.java
@@ -1,0 +1,56 @@
+/*
+PatternVariableName
+format = (default)^([a-z][a-zA-Z0-9]*|_)$
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.naming.patternvariablename;
+
+public class InputPatternVariableNameUnnamed {
+
+    void test(Object o, Object obj) {
+        if (o instanceof String _) { }
+        if (o instanceof String) { }
+        // violation below, 'Name '__' must match pattern*.'
+        if (o instanceof String __ && __.length() == 0) { }
+        // violation below, 'Name '_s' must match pattern*.'
+        if (o instanceof String _s) { }
+
+        // violation below, 'Name '__' must match pattern*.'
+        if (o instanceof String __ &&
+                        obj instanceof Integer _) { }
+    }
+
+    void test2(Object o) {
+        switch (o) {
+            // violation below, 'Name '__' must match pattern*.'
+            case String __ when __.length() == 0 -> {}
+            case Integer _ -> {}
+            // violation below, 'Name '_s' must match pattern*.'
+            case Double _s -> {}
+            default -> {}
+        }
+    }
+
+    void test3(Object o, Object obj) {
+        if (o instanceof Point(int _, int _)) { }
+          // violation below, 'Name '_Color' must match pattern*.'
+        if (o instanceof ColoredPoint(Point(int _, int x), String _Color)) { }
+
+        switch (o) {
+            case Point(int _, int _) -> {}
+              // violation below, 'Name '_Color' must match pattern*.'
+            case ColoredPoint(Point(int _, int x), String _Color) -> {}
+            default -> {}
+        }
+
+        boolean b = o instanceof Point(int _, int y)
+                  // violation below, 'Name '__' must match pattern*.'
+                && obj instanceof ColoredPoint(Point(int _, int x), String __);
+    }
+
+    record Point(int x, int y) { }
+    record ColoredPoint(Point point, String color) { }
+}

--- a/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckExamplesTest.java
+++ b/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckExamplesTest.java
@@ -34,7 +34,7 @@ public class PatternVariableNameCheckExamplesTest extends AbstractExamplesModule
     @Test
     public void testExample1() throws Exception {
 
-        final String pattern = "^[a-z][a-zA-Z0-9]*$";
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
 
         final String[] expected = {
             "15:30: " + getCheckMessage(MSG_INVALID_PATTERN, "STRING", pattern),

--- a/src/xdocs/checks/naming/patternvariablename.xml
+++ b/src/xdocs/checks/naming/patternvariablename.xml
@@ -28,7 +28,7 @@
               <td>format</td>
               <td>Sets the pattern to match valid identifiers.</td>
               <td><a href="../../property_types.html#Pattern">Pattern</a></td>
-              <td><code>&quot;^[a-z][a-zA-Z0-9]*$&quot;</code></td>
+              <td><code>&quot;^([a-z][a-zA-Z0-9]*|_)$&quot;</code></td>
               <td>8.36</td>
             </tr>
           </table>

